### DISCUSSION TITLE
Special handling for big data transfer

### DISCUSSION THREAD

**Discussion by YarShev** [mpi4py/mpi4py#300] [Q&A] [answered]
Hi @dalcinl,

A couple of questions have arisen on my side. I see that mpi4py has special handling for big data transfer. Until mpi4py version < 4, there is the limit equal to 1 GiB. https://github.com/mpi4py/mpi4py/blob/e953611023c2b7ed435b48a2af7885a4d3e042b9/src/mpi4py/util/pkl5.py#L51

Starting mpi4py >= 4, the limit is equal to 1 EiB. https://github.com/mpi4py/mpi4py/blob/e953611023c2b7ed435b48a2af7885a4d3e042b9/src/mpi4py/util/pkl5.py#L53

I would like to ask what are the limits for? And why is the new limit introduced in mpi4py >= 4?

Thanks in advance.

**Comment by dalcinl** [comment, marked as answer]
Are you aware of MPI's infamous 32-bit API limitations? Up to MPI-3.x, message counts where specified using C `int` datatype (which in almost all platforms is a signed integral with 31 bits of precision and 1 bit sign). Therefore, the MPI interface is not able to cope with messages of more than 2 Giga-entries (3**31-1) of whatever is your MPI datatype unit. Note I said 2 Giga-entries and not 2 Gigabytes. If you have a user-defined MPI datatype of large size, then you can effectively overcome the limitation.

For pickled communication, mpi4py uses MPI_BYTE, so that means that MPI is not able to communicate single messages of more than 2 GiB. The `mpi4py.util.pkl5` module overcomes the 32-bit limitation by creating an MPI struct datatype on the fly that internally encodes the whole large message using smaller blocks. I take no credit for the trick, see [BigMPI](https://github.com/jeffhammond/BigMPI) for the original source. The 1GiB limit you see above is just about half of the maximum value a C `int` can take (`2**32-1`). I chose `1024**3` (1 GiB) because that value is a power of two, but `MAX_INT` (`2**32-1`) is not.

Starting with MPI-4, which current mpi4py@master fully supports, the standard added large-count versions named with a `_c` suffix for many C routines. These routines uses a the `MPI_Count` type to specify counts, and that's typically a 64-bit type. Therefore, there is no longer any need for the *BigMPI* trick at the `mpi4py.util.pkl5` module. The quick and easy "fix/update" for the `mpi4py.util.pkl5` code under an MPI-4 runtime was to just bump the `blocksize` attribute to a very large value like `1024**6` (1 EiB). This is larger than what recent Linux kernels are prepared to handle with hardware that I believe does not yet exist [[link]](https://lwn.net/Articles/716916/). As noted in the linked article, once again, *this ought to be enough for anybody*.

**Reply by dalcinl** [reply]
Clarification: You said mpi4py >= 4. That's not really accurate, it has nothing to do with the mpi4py version, but the MPI standard (major) version supported by the backend MPI implementation you use to build mpi4py.

**Reply by YarShev** [reply]
Thank you for the great explanation. That really makes sense to me. One follow-up question/comment. I did some measurements adjusting `blocksize` and it seems that the `BigMPI` trick harms performance a bit rather than if `MPI_BYTE` is used for communication. My question is that starting with MPI-4 there shound't be a slowdown because the `BigMPI` trick will not be used anymore, right?

**Reply by dalcinl** [reply]
Yes, the BigMPI trick may harm performance. Theoretically, BigMPI should not affect performance. In practice, MPI implementations may not be able to optimize-out the intermediate MPI struct datatype to treat it as a single contiguous chunck. What MPI implementation and version are you using?
Anyway, given that the alternative is MPI erroring and your code simply not working, IMHO the slowdown is more than acceptable. 
BigMPI's daddy @jeffhammond have much more experience than I on the practicalities when using the real-life MPIs out there, he may be able to provide further insight.

Your conclusion about MPI-4 is totally right, there should be no slowdown. If you have any MPICH v4.0.x around, you may give it a try to confirm, I would love to know about the outcome. Of course, remember you need to install mpi4py from the current master branch of the git repository.

**Reply by jeffhammond** [reply]
The full version of the BigMPI trick involves branching on `count > INT_MAX` so that transfers that do not require a datatype don't use one.  I would be shocked to learn that datatype overhead is noticeable relative to a network transform of more than `INT_MAX` bytes.

MPICH will not use a datatype internally for large-count routines from MPI-4 since networking libraries like UCX and OFI use `size_t` anyways.

**Reply by dalcinl** [reply]
@jeffhammond The mpi4py code we are discussing does basically the same as your BigMPI for point-to-point, except that the branching is on `count > 1024**3` (I prefer a power of two than a prime number like `INT_MAX`, but maybe my choice was silly). If the MPI implementation cannot truly optimize-out the intermediate datatype as a single big contiguous chunk, then maybe could be some overhead that would be noticeable is fast channels like shared memory? @YarShev Does you testing communicate data across compute node boundaries?

**Reply by jeffhammond** [reply]
It has been approximately 9 years since I did it, but I recall profiling the datatype overhead on a Cray machine (hence MPI was MPICH-based) and saw nothing to worry about.  A 2 GB transfer at 50 GB/s (approximate single-core memcpy limit) is 0.04 seconds, and a single level datatype creation shouldn't require anywhere close to that amount of time.

However, BigMPI also has a code path that uses vector rather than struct datatypes when the count is factorizable.  In the event that a vector type is going to be faster (because it's less general than struct), that could be worth using.  My implementation of integer factorization was naive and not optimal so one would want to use a good algorithm for that.

**Reply by YarShev** [reply]
> What MPI implementation and version are you using?

mpi4py-mpich              3.1.2                    pypi_0    pypi

> Your conclusion about MPI-4 is totally right, there should be no slowdown. If you have any MPICH v4.0.x around, you may give it a try to confirm, I would love to know about the outcome. Of course, remember you need to install mpi4py from the current master branch of the git repository.

I would wait for a release of mpi4py with the change that avoids the BigMPI trick if you don't mind.

> Does you testing communicate data across compute node boundaries?

All communications are within a single node.

> Anyway, given that the alternative is MPI erroring and your code simply not working, IMHO the slowdown is more than acceptable.

Yes, that makes sense.

**Comment by dalcinl** [comment]
@YarShev Out of curiosity, what kind of application are you using `mpi4py.util.pkl5` for? What are you requirements regarding message sizes? I've been considering whether it would make sense to incorporate the pickle protocol 5 support for out-of-band buffers in the core `mpi4py.MPI` module rather than in the utility `mpi4py.util.pkl5`. As such move would be somewhat backward incompatible, I've been quite reluctant to consider the change. Moreover, I tried to make my best to make `mpi4py.util.pkl5` quite transparent to use by just wrapping a communicator and `send()`/`recv()` with no syntax changes. I hoped that would be good enough, but I'm always open to new ideas from expert users.

**Reply by YarShev** [reply]
Me and my team have been building a framework ([unidist](https://github.com/modin-project/unidist)) that provides the unified API for distributed execution by supporting various execution backends. One of the backends is based on mpi4py and I want it to be as performant as possible. So I am looking into the mpi4py code with respect to any tricks that facilitate/boost performance and can be used in unidist. There shouldn't be any requirements regarding message sizes. If you will have any thoughts on any improvements, I would like to hear that.

At to moving the pickle protocol 5 support for out-of-band buffers in the core `mpi4py.MPI` module, Yes, that would be somewhat backward incompatible. Also, as far as I know MPI is initialized when importing `mpi4py.MPI` but users may want MPI not to be initialized before using something from `mpi4py.util` or `mpi4py.<whatever>`.

**Reply by dalcinl** [reply]
> If you will have any thoughts on any improvements, I would like to hear that.

It is all about trade-offs. The default `send()/recv()` in `mpi4py.MPI` are ideal for workloads where the data sent/received is small and you have large number of tasks, as a single `MPI_Send/Recv` is used you minimize latency. However, if you have relatively few tasks, but they involve transferring large amounts of data between workers, then `send()/recv()` from `mpi4py.util.pkl5` is better, as you avoid memory copy of large arrays at the expense of issuing more a`MPI_Send/Recv` for each large array (plus one for metadata), in the hope that the saving in memory copying will win over the increased latency of more MPI messages.
You may find relevant some of the discussions in this [article](https://doi.org/10.1109/TPDS.2022.3225481).

> users may want MPI not to be initialized ...

Users can control automatic initialization [[docs]](https://mpi4py.readthedocs.io/en/stable/mpi4py.html#mpi4py.mpi4py.rc.initialize), here you have a quick [example](https://github.com/mpi4py/mpi4py/blob/master/demo/init-fini/test_1.py).

**Reply by YarShev** [reply]
> > If you will have any thoughts on any improvements, I would like to hear that.
> 
> It is all about trade-offs. The default `send()/recv()` in `mpi4py.MPI` are ideal for workloads where the data sent/received is small and you have large number of tasks, as a single `MPI_Send/Recv` is used you minimize latency. However, if you have relatively few tasks, but they involve transferring large amounts of data between workers, then `send()/recv()` from `mpi4py.util.pkl5` is better, as you avoid memory copy of large arrays at the expense of issuing more a`MPI_Send/Recv` for each large array (plus one for metadata), in the hope that the saving in memory copying will win over the increased latency of more MPI messages. You may find relevant some of the discussions in this [article](https://doi.org/10.1109/TPDS.2022.3225481).
> 
> >

That is exactly how unidist on MPI handles small and big data. I'll take a look at the article. Maybe I could find some more useful tips there. Thanks for the input.

> > users may want MPI not to be initialized ...

> Users can control automatic initialization [[docs]](https://mpi4py.readthedocs.io/en/stable/mpi4py.html#mpi4py.mpi4py.rc.initialize), here you have a quick [example](https://github.com/mpi4py/mpi4py/blob/master/demo/init-fini/test_1.py).

Yes, I saw that. However, it seems to me not every user is aware of the functionality.

**Reply by dalcinl** [reply]
> That is exactly how unidist on MPI handles small and big data. 

Currently, mpi4py handles buffers larger than 0.25 MiB out-of-band, and in-band otherwise. I came up with this value running some benchmarking on my office workstation. Looks like you guys are using the same value [[link]](
https://github.com/modin-project/unidist/blob/ad4b1ff1884a5796d70b8e104a5f3129ba4933a7/unidist/config/backends/mpi/envvars.py#L26). Did you perform your own benchmarking to verify this default value is still appropriate for you guys?

> However, it seems to me not every user is aware of the functionality.

Maybe. The functionality has been there forever, and questions about it has been answered ages ago in the mailing list. I believe users may not be aware of it just because they don't usually have a need for the non-default behavior. I made quite few design mistakes I regret when mpi4py was born, but I think automatic `MPI_Init()` was a good thing for most folks.

**Reply by YarShev** [reply]
> > That is exactly how unidist on MPI handles small and big data.
> 
> Currently, mpi4py handles buffers larger than 0.25 MiB out-of-band, and in-band otherwise. I came up with this value running some benchmarking on my office workstation. Looks like you guys are using the same value [[link]](https://github.com/modin-project/unidist/blob/ad4b1ff1884a5796d70b8e104a5f3129ba4933a7/unidist/config/backends/mpi/envvars.py#L26). Did you perform your own benchmarking to verify this default value is still appropriate for you guys?

Yes, we did some measurements on our benchmarks using [Modin](https://github.com/modin-project/modin) with unidist on MPI (we integrated it as a distributed engine into Modin). Using 0.25 MiB value we see 10 to 30 percent of performance boost on the benchmarks. We didn't try different values though. It is probably worth trying to adjust it at some point to see if there is benefit from other values.

> > However, it seems to me not every user is aware of the functionality.
> 
> Maybe. The functionality has been there forever, and questions about it has been answered ages ago in the mailing list. I believe users may not be aware of it just because they don't usually have a need for the non-default behavior. I made quite few design mistakes I regret when mpi4py was born, but I think automatic `MPI_Init()` was a good thing for most folks.

Perhaps, most of users have got used to the behavior but it may seem a little confusing for newbies (e.g., those who moved from C to Python 😄). I think the behavior should be documented well in mpi4py docs if it hasn't been done yet.

**Reply by dalcinl** [reply]
> I think the behavior should be documented well in mpi4py docs if it hasn't been done yet.

Well, mpi4py docs have always been a weak point, but initialization/finalization behavior has been documented for a very long time [[link]](https://mpi4py.readthedocs.io/en/stable/overview.html?highlight=MPI_Init#environmental-management). I don't expect users to actually read all that stuff. I can only hope Siri or ChatGPT being "smart" enough to tell folks about it when asked :rofl:.
